sql/ttl: centralize row-level TTL progress updates at coordinator

spilchen opened the pull request:
Previously, each processor in a row-level TTL job directly updated job progress, leading to fan-out issues in large clusters and performance hits during restarts when many spans had no expired rows. Gating logic was previously added to mitigate this.

This change delegates all progress updates to the coordinator. Each processor now sends its progress to the coordinator, which performs the job progress update. The original gating logic remains but is now enforced by the coordinator.

Closes: #135227

Epic: none
Release note: none

rafiss reviewed (changes requested):
great work! i mostly have minor comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-:-OTmXylYAjcKXLS4qThs:b-pkpk6w)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @kev-cao)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 369 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OTmS-R96CVWCFqHvM4Q:-OTmS-R96CVWCFqHvM4R:bw8j1dk) ([raw file](https://github.com/cockroachdb/cockroach/blob/2977618130ce76003edf75c29f11dff7bb438d43/pkg/sql/ttl/ttljob/ttljob.go#L369)):*
> ```Go
> // initProgressInJob initializes and persists the initial RowLevelTTL job progress state.
> // It computes throttling thresholds and writes an empty progress object to the job record.
> // This should be called once before job execution starts.
> ```

This is not something to change in this PR, but just confirming: once we address https://github.com/cockroachdb/cockroach/issues/140514, then we would want to remove this initialization logic, correct?
___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 466 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OTmUgbA8yn0qtNvMmjr:-OTmUgbA8yn0qtNvMmjs:bryztgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2977618130ce76003edf75c29f11dff7bb438d43/pkg/sql/ttl/ttljob/ttljob.go#L466)):*
> ```Go
> 		rowLevelTTL.ProcessorProgresses = append(rowLevelTTL.ProcessorProgresses, incomingProcProgress)
> 		rowLevelTTL.JobDeletedRowCount += incomingProcProgress.DeletedRowCount
> 		rowLevelTTL.JobProcessedSpanCount += incomingProcProgress.ProcessedSpanCount
> ```

I'm not sure I follow the counting logic here -- it seems like we are incrementing `JobDeletedRowCount`, `JobProcessedSpanCount`, and `totalSpanCount` twice in the case where `existingProcProgress.ProcessorID` is not equal to `incomingProcProgress.ProcessorID`
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OTmWZZb8_M1ZNoD5XMh:-OTmWZZb8_M1ZNoD5XMi:b-xzq911) ([raw file](https://github.com/cockroachdb/cockroach/blob/2977618130ce76003edf75c29f11dff7bb438d43/pkg/sql/ttl/ttljob/ttljob_processor.go#L72)):*
> ```Go
> func (t *ttlProcessor) Start(ctx context.Context) {}
> 
> func (t *ttlProcessor) Run(ctx context.Context, output execinfra.RowReceiver) {
> ```

nit: a comment like "// Run implements the execinfra.Processor interface." could be useful. let's also comment `Start` while we're at it
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OTmX_2UEM1ahEiYy_sZ:-OTmX_2VBtcnu7jLxhqn:b-315exz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2977618130ce76003edf75c29f11dff7bb438d43/pkg/sql/ttl/ttljob/ttljob_processor.go#L321)):*
> ```Go
> 	status := output.Push(nil, meta)
> 	if status != execinfra.NeedMoreRows {
> 		return errors.Errorf("output receiver rejected progress metadata: %v", status)
> ```

could this happen in a mixed-version cluster?
___
*[`pkg/jobs/jobspb/jobs.proto` line 1368 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OTmUAFw6tWGtw2q4SjL:-OTmUAFw6tWGtw2q4SjM:b-puznw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2977618130ce76003edf75c29f11dff7bb438d43/pkg/jobs/jobspb/jobs.proto#L1368)):*
> ```proto
>   int64 deleted_row_count = 3;
> 
>   // SpanCount is the total number of spans of the DistSQL processor;
> ```

nit: the comment doesn't match the variable name. also, perhaps it would be more clear to say "total number of spans assigned to the DistSQL processor."


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148914)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @kev-cao, and @spilchen)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OTmYgzSDRXJHE2KlOsG:-OTmYgzSDRXJHE2KlOsH:b1jzmka):*
What "gating logic" does this refer to?
___
*[`-- commits` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OTmYqQlB3M4ZwS62Tt1:-OTmYqQlB3M4ZwS62Tt2:b-i5j9it):*
I've only skimmed the PR, but I think we need to add special handling for mixed-version clusters.

Namely, 25.2 coordinator doesn't expect to receive any progress updates from newer remote nodes, so we need to add a version gate. `execversion.V` is probably needed - the gateway node of TTL job would pick the exec version based on the cluster version - we'll need to add a new one for 25.3 and adjust `DistSQLPlanner.setupFlows` to use it once the cluster version is at least 25.3. Then each TTL processor running new binary would determine whether 25.2 behavior is needed, in which case it'd proceed to update progress itself, otherwise it can send it as metadata.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148914)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @kev-cao, @rafiss, and @yuzefovich)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OTmYgzSDRXJHE2KlOsG:-OTmkwj48GnUOtqF62mG:bolzncx):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

What "gating logic" does this refer to?
</blockquote></details>

This was the stuff in ttljob_processor.go that tried to slow down updates to jobs progress:
https://github.com/cockroachdb/cockroach/blob/c0ced8a60eca9a55e714e8e04b8feb950156d190/pkg/sql/ttl/ttljob/ttljob_processor.go#L188-L192

That no longer exists in the processor and is handled by the coordinator now.
___
*[`-- commits` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OTmYqQlB3M4ZwS62Tt1:-OTmyVPJ86CDyV1FBLcn:b-8fpgix):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I've only skimmed the PR, but I think we need to add special handling for mixed-version clusters.

Namely, 25.2 coordinator doesn't expect to receive any progress updates from newer remote nodes, so we need to add a version gate. `execversion.V` is probably needed - the gateway node of TTL job would pick the exec version based on the cluster version - we'll need to add a new one for 25.3 and adjust `DistSQLPlanner.setupFlows` to use it once the cluster version is at least 25.3. Then each TTL processor running new binary would determine whether 25.2 behavior is needed, in which case it'd proceed to update progress itself, otherwise it can send it as metadata.
</blockquote></details>

Ah yes, good catch. I didn't consider mixed-version. This change won't be back ported to 25.3. So I ended up using 25.4 as the new distSQL flow version.
___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 369 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OTmS-R96CVWCFqHvM4Q:-OTmmAMa8Ux9LwkD5LHC:bamq38f) ([raw file](https://github.com/cockroachdb/cockroach/blob/2977618130ce76003edf75c29f11dff7bb438d43/pkg/sql/ttl/ttljob/ttljob.go#L369)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

This is not something to change in this PR, but just confirming: once we address https://github.com/cockroachdb/cockroach/issues/140514, then we would want to remove this initialization logic, correct?
</blockquote></details>

Yes, the checkpointing work will overhaul the init logic. Much of it can go, especially the gating logic that slows update frequency since that becomes unnecessary. We also won’t need to reset the processed span count to zero, since retries won’t start from the beginning. We’ll likely still need to track the total span count, but only on the job’s first run.
___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 466 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OTmUgbA8yn0qtNvMmjr:-OTmptO-EM8ZBuYkUca2:b-v9kspz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2977618130ce76003edf75c29f11dff7bb438d43/pkg/sql/ttl/ttljob/ttljob.go#L466)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

I'm not sure I follow the counting logic here -- it seems like we are incrementing `JobDeletedRowCount`, `JobProcessedSpanCount`, and `totalSpanCount` twice in the case where `existingProcProgress.ProcessorID` is not equal to `incomingProcProgress.ProcessorID`
</blockquote></details>

Hmm. I think the current logic is correct. If `foundMatchingProcessor` is false, it means the processor wasn’t in `ProcessorProgresses`, so it’s correct to append it and then include its counts.

That said, the code is doing two things in one loop: updating the processor list and summing the job-level counters. That makes it harder to reason about. I'll split those steps. First, update or insert the incoming processor. Then, in a separate loop, recompute the job-level totals. I think that will make the logic clearer.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OTmX_2UEM1ahEiYy_sZ:-OU4hcKs33TkHV2sv1_z:b-erdfhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2977618130ce76003edf75c29f11dff7bb438d43/pkg/sql/ttl/ttljob/ttljob_processor.go#L321)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could this happen in a mixed-version cluster?
</blockquote></details>

Good point about mixed versions. I didn't properly handle it. We now support both job progress update types in TTL. We pick the correct method based on the dist SQL version.

This brought in additional test changes, as I want to run the existing TTL unit tests with the old progress updater.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Nice work! DistSQL changes LGTM.

> Reviewed 1 of 9 files at r1, 4 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-:-OU7b7E5EW9fx_oA0byg:b-xbcw15)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @kev-cao, @rafiss, and @spilchen)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OTmYgzSDRXJHE2KlOsG:-OU7U_lM7tliJBKc9lzY:be17u8d):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

This was the stuff in ttljob_processor.go that tried to slow down updates to jobs progress:
https://github.com/cockroachdb/cockroach/blob/c0ced8a60eca9a55e714e8e04b8feb950156d190/pkg/sql/ttl/ttljob/ttljob_processor.go#L188-L192

That no longer exists in the processor and is handled by the coordinator now.
</blockquote></details>

Ack, thanks.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 159 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OU7YeVO1PgOwR1D4DVk:-OU7YeVO1PgOwR1D4DVl:b-ovqdqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0f7c1dad1a823c0abea267c79dd1dd3078aefcd/pkg/sql/ttl/ttljob/ttljob_processor.go#L159)):*
> ```Go
> 		output.Push(nil, &execinfrapb.ProducerMetadata{Err: err})
> 	}
> 	output.ProducerDone()
> ```

nit: not sure whether we care about this, but we might want to add `execinfra.SendTraceData` call before doing `output.ProducerDone()` - this will include the trace information that is captured in the span (which is created in `t.StartInternal`) and will propagate it as ProducerMetadata to the gateway (the coordinator on the gateway will inject it into its own tracing span which would allow us to view the distributed trace). However, I'm not sure whether we actually expose that distributed trace in any way, so it might be simply ignored on the gateway node.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 561 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OU7_taCAel3FGyLiUW3:-OU7_taCAel3FGyLiUW4:b-dziwa8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0f7c1dad1a823c0abea267c79dd1dd3078aefcd/pkg/sql/ttl/ttljob/ttljob_processor.go#L561)):*
> ```Go
> 		},
> 	}
> 	// Push progress after each span. Throttling is now handled by the coordinator.
> ```

How quickly does a TTL processor handles a single span and how many spans does each processor handle? I'm just wondering if pushing a metadata object (which is likely to be sent across network at DistSQL level) is too frequent or not. If it's like once per second or less frequent, then it's fine, but if it can be 100 spans per second per processor, it might be too much (?).


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148914)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @kev-cao, @rafiss, and @yuzefovich)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 561 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OU7_taCAel3FGyLiUW3:-OU9yeBF8vdFAM5bD75f:b1lzvz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0f7c1dad1a823c0abea267c79dd1dd3078aefcd/pkg/sql/ttl/ttljob/ttljob_processor.go#L561)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

How quickly does a TTL processor handles a single span and how many spans does each processor handle? I'm just wondering if pushing a metadata object (which is likely to be sent across network at DistSQL level) is too frequent or not. If it's like once per second or less frequent, then it's fine, but if it can be 100 spans per second per processor, it might be too much (?).
</blockquote></details>

The common case it will be less frequent. However, there isn't any checkpointing (yet) so if the job needs to retry it can easily burn through quite a few spans (likely sub-second) as the spans that had already been processed wouldn't have any deleted data. Do you think that warrants having the gating logic in the processor still? At least until we add the checkpointing (#140514)


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148914)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @kev-cao, @rafiss, and @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 561 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OU7_taCAel3FGyLiUW3:-OUB4gZa-zxA3awlSZeQ:b96b1z3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0f7c1dad1a823c0abea267c79dd1dd3078aefcd/pkg/sql/ttl/ttljob/ttljob_processor.go#L561)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

The common case it will be less frequent. However, there isn't any checkpointing (yet) so if the job needs to retry it can easily burn through quite a few spans (likely sub-second) as the spans that had already been processed wouldn't have any deleted data. Do you think that warrants having the gating logic in the processor still? At least until we add the checkpointing (#140514)
</blockquote></details>

Probably not necessary then. We do send millions of rows in DistSQL plans between processors, so if we happen to send a few thousands of metadata objects on a TTL retry it's probably a drop in the bucket.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
nice work, this looks really close! my comments are all pretty narrow

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-:-OUel89bARqjkZJP65fb:b-7m1ft7)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @kev-cao, and @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 77 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OUehpam3Sgx3d9u-qAj:-OUehpam3Sgx3d9u-qAk:b-amvri8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8cacbf7f0fd2077edcd1d431aad3ca4984c1d4c/pkg/sql/ttl/ttljob/ttljob_processor.go#L77)):*
> ```Go
> // metadata back to the coordinator (preferred).
> type ttlProgressUpdater interface {
> 	// InitProgress is called once at the beginning of the TTL processor
> ```

super nit: add periods to the ends of the sentences.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 144 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OUeiIqu3O5zZFuhBb_I:-OUeiIqu3O5zZFuhBb_J:b8ony4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8cacbf7f0fd2077edcd1d431aad3ca4984c1d4c/pkg/sql/ttl/ttljob/ttljob_processor.go#L144)):*
> ```Go
> func (t *ttlProcessor) Run(ctx context.Context, output execinfra.RowReceiver) {
> 	ctx = t.StartInternal(ctx, "ttl")
> 	v := execversion.FromContext(ctx)
> ```

nice, I didn't know about `execversion.FromContext`, but it seems pretty handy. do we know when we can rely upon it? asking because it seems like the behavior if the version key is not present is to panic, which could be disruptive.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 150 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OUeiY_L4pWioa2Oq7cg:-OUeiY_L4pWioa2Oq7ch:b-crd60e) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8cacbf7f0fd2077edcd1d431aad3ca4984c1d4c/pkg/sql/ttl/ttljob/ttljob_processor.go#L150)):*
> ```Go
> 	// The selected behavior is gated on the active cluster version.
> 	// TODO(spilchen): remove directJobProgerssUpdater once 25.4 is the minimum supported version.
> 	if v == execversion.V25_2 {
> ```

this seems odd to me -- at this point in time, shouldn't there be a `V25_3` version? it seems like the DistSQL version is a bit different than the clusterversion checks that we use in most other places in the code. perhaps @yuzefovich could help us understand.

regardless, i wonder if we need to use less-than or greater-than comparisons for the versions rather than equality. that way we can ensure we fallback to the old behavior for all versions that are old enough.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 525 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OUejZrf7YfPscq6SHxp:-OUejZrf7YfPscq6SHxq:b-pccvhj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8cacbf7f0fd2077edcd1d431aad3ca4984c1d4c/pkg/sql/ttl/ttljob/ttljob_processor.go#L525)):*
> ```Go
> }
> 
> // InitProgress implements the ttlProgressUpdater interface
> ```

super nit: add periods at the end of these function comments.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 629 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OUejxbt8E8t48F3cPzz:-OUejxbt8E8t48F3cQ--:bvjt2gf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8cacbf7f0fd2077edcd1d431aad3ca4984c1d4c/pkg/sql/ttl/ttljob/ttljob_processor.go#L629)):*
> ```Go
> 		return err
> 	}
> 	log.Infof(
> ```

this logging would be useful to preserve for the coordinatorStreamUpdater implementation also. i think we can keep the behavior of each node logging something like this before it pushes progress back to the coordinator. that would be useful to have in case something goes wrong with the RPC calls to send the progress objects between different nodes.

that being said, we probably still do want to have rate-limiting on this logging call in the coordinatorStreamUpdater implementation too. if there's not a natural place to rate-limit in the processor, then we can just do the logging in the coordinator.
___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 384 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OUehMV0DXwbs25WZnHo:-OUehMV0DXwbs25WZnHp:b-8lszz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8cacbf7f0fd2077edcd1d431aad3ca4984c1d4c/pkg/sql/ttl/ttljob/ttljob.go#L384)):*
> ```Go
> // from the last producer metadata received.
> //
> // In mixed-version clusters, TTL processors fall back to direct job table updates
> ```

super nit: could this comment specify "mixed version clusters from 25.3 and earlier" or something to that effect? that way we can easily see when this comment goes out of date.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148914)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @kev-cao, @rafiss, and @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 144 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OUeiIqu3O5zZFuhBb_I:-OUf8niBE9pw1IrvuLIy:b8y54wi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8cacbf7f0fd2077edcd1d431aad3ca4984c1d4c/pkg/sql/ttl/ttljob/ttljob_processor.go#L144)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nice, I didn't know about `execversion.FromContext`, but it seems pretty handy. do we know when we can rely upon it? asking because it seems like the behavior if the version key is not present is to panic, which could be disruptive.
</blockquote></details>

The version is included in `setupFlow` into the context object that we consider the "flow" context, which all goroutines should be using (either directly or its descendants). Things would break only if we used `context.Background()` or something like that somewhere, but we don't in the distsql infra, so we should be good with the current code. On Queries team we've done a couple of migrations in the stats collection code using `execversion.V` mechanism (#138561, #140601), and we haven't seen any issues.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 150 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OUeiY_L4pWioa2Oq7cg:-OUf6hjY5cP-B_u3h9Qf:b-k3igjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8cacbf7f0fd2077edcd1d431aad3ca4984c1d4c/pkg/sql/ttl/ttljob/ttljob_processor.go#L150)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this seems odd to me -- at this point in time, shouldn't there be a `V25_3` version? it seems like the DistSQL version is a bit different than the clusterversion checks that we use in most other places in the code. perhaps @yuzefovich could help us understand.

regardless, i wonder if we need to use less-than or greater-than comparisons for the versions rather than equality. that way we can ensure we fallback to the old behavior for all versions that are old enough.
</blockquote></details>

We don't need to bump `execversion.V` for every CRDB version, we only need to do this whenever we introduce a change in the behavior of existing processors that need to be gated based on this version, so `execversion.V25_2` is a valid predecessor of `V25_4`.

We could make this condition to be an inequality, but it's not needed. Since all CRDB binaries right now need to be compatible with CRDB version 25.2+, distsql code will actually reject any flows that have `execversion.V25_1` or prior (search for "version mismatch in flow request").

For some more context, this `execversion.V` mechanism allows us to pick "execution behavior" at a distsql plan granularity, meaning that all nodes might be running newer binaries, but the gateway node could pick the "old" execution behavior. Cluster version mechanism doesn't support such things since each node in the cluster could learn about the version upgrade at a different point in time, so nodes within one distsql plan could have different cluster versions at the time of flow setup.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148914)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @kev-cao, @rafiss, and @yuzefovich)

___
*[`pkg/sql/ttl/ttljob/ttljob.go` line 384 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OUehMV0DXwbs25WZnHo:-OUixUKU6mUPfIvF4JcQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8cacbf7f0fd2077edcd1d431aad3ca4984c1d4c/pkg/sql/ttl/ttljob/ttljob.go#L384)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: could this comment specify "mixed version clusters from 25.3 and earlier" or something to that effect? that way we can easily see when this comment goes out of date.
</blockquote></details>

Done.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 77 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OUehpam3Sgx3d9u-qAj:-OUixsxx1HtC4El040u9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8cacbf7f0fd2077edcd1d431aad3ca4984c1d4c/pkg/sql/ttl/ttljob/ttljob_processor.go#L77)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: add periods to the ends of the sentences.
</blockquote></details>

Done.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 150 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OUeiY_L4pWioa2Oq7cg:-OUiykVXETJDa1syiYLy:b90w2ku) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8cacbf7f0fd2077edcd1d431aad3ca4984c1d4c/pkg/sql/ttl/ttljob/ttljob_processor.go#L150)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

We don't need to bump `execversion.V` for every CRDB version, we only need to do this whenever we introduce a change in the behavior of existing processors that need to be gated based on this version, so `execversion.V25_2` is a valid predecessor of `V25_4`.

We could make this condition to be an inequality, but it's not needed. Since all CRDB binaries right now need to be compatible with CRDB version 25.2+, distsql code will actually reject any flows that have `execversion.V25_1` or prior (search for "version mismatch in flow request").

For some more context, this `execversion.V` mechanism allows us to pick "execution behavior" at a distsql plan granularity, meaning that all nodes might be running newer binaries, but the gateway node could pick the "old" execution behavior. Cluster version mechanism doesn't support such things since each node in the cluster could learn about the version upgrade at a different point in time, so nodes within one distsql plan could have different cluster versions at the time of flow setup.
</blockquote></details>

Okay, I think we're good with the change. Yahor noted that this could be an inequality, but I'll leave it as-is. If another DistSQL version is introduced, we want it to take the else path and continue using the new progress code.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 525 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OUejZrf7YfPscq6SHxp:-OUj-lO-ELw_MjiDpTd3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8cacbf7f0fd2077edcd1d431aad3ca4984c1d4c/pkg/sql/ttl/ttljob/ttljob_processor.go#L525)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: add periods at the end of these function comments.
</blockquote></details>

Done.
___
*[`pkg/sql/ttl/ttljob/ttljob_processor.go` line 629 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/148914#-OUejxbt8E8t48F3cPzz:-OUj4sUu5pJbY73yrRjI:bu5l3o1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8cacbf7f0fd2077edcd1d431aad3ca4984c1d4c/pkg/sql/ttl/ttljob/ttljob_processor.go#L629)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this logging would be useful to preserve for the coordinatorStreamUpdater implementation also. i think we can keep the behavior of each node logging something like this before it pushes progress back to the coordinator. that would be useful to have in case something goes wrong with the RPC calls to send the progress objects between different nodes.

that being said, we probably still do want to have rate-limiting on this logging call in the coordinatorStreamUpdater implementation too. if there's not a natural place to rate-limit in the processor, then we can just do the logging in the coordinator.
</blockquote></details>

Good idea. I added logging and used a `log.EveryN` struct to control how frequent to log. It will log about once every minute.


<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/ttl/ttljob/ttljob_processor.go:
```diff
@@ -522,6 +526,176 @@ func newTTLProcessor(
+		return errors.Errorf("output receiver rejected progress metadata: %v", status)
+	}
+	if c.progressLogger.ShouldLog() {
+		log.Infof(ctx, "TTL processor progress: %+v", progressMsg)
```
to avoid this getting redacted in the logs, i think we need to make `RowLevelTTLProcessorProgress` implement `redact.SafeFormatter` (and use `%v` here)

spilchen replied:
Done.
<!-- Reviewable comment -OUuu5Ov4FlSs3YdVYLp:b-896fix -->
<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm! i just have one more comment about logging -- though if you tested it and i'm wrong, feel free to ignore that.

spilchen committed:
sql/ttl: centralize row-level TTL progress updates at coordinator

Previously, each processor in a row-level TTL job directly updated job
progress, leading to fan-out issues in large clusters and performance
hits during restarts when many spans had no expired rows. Gating logic
was previously added to mitigate this.

This change delegates all progress updates to the coordinator. Each
processor now sends its progress to the coordinator, which performs the
job progress update. The original gating logic remains but is now
enforced by the coordinator.

Closes: #135227

Epic: none
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148914)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @kev-cao, @rafiss, and @yuzefovich)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTRs!

bors r+

rafiss commented:
bors retry
